### ISSUE TITLE
PHP error: "failed to open stream: Too many open files"

### ISSUE THREAD

**Issue by TKafassis** [guzzle/guzzle#1927]
I tried to do like 5k concurrent requests and got the above error. Can somebody help me please?

**Comment by alexeyshockov** [comment]
Try to increase open file limit (depends on your OS, for Linux read [this manual](https://www.tecmint.com/increase-set-open-file-limits-in-linux/), for example).

But I would advice to decrease concurrent connections for a process.

**Comment by TKafassis** [comment]
Thanks for the fast response! By the way, before I was able to do that with multi curl without any such problems, how can that be explained? I'm sorry but i'm just starting to understand and use Guzzle.

**Comment by sagikazarmark** [comment]
Guzzle writes the body to streams, which in turn are temp files. That's why it worked with curl for you.

**Comment by skatetdieu** [comment]
Hello, I keep getting this error as well when using this [PHP Web Push notifications library](https://github.com/web-push-libs/web-push-php) in a loop to send out notifications to thousands of URL API end points.

How would I go about decreasing the concurrent connections while processing as suggested?

Thank you in advance for any help.

**Comment by sagikazarmark** [comment]
I guess it depends on the library. Taking a quick look at the library: https://github.com/web-push-libs/web-push-php/blob/master/src/WebPush.php#L136

**Comment by skatetdieu** [comment]
Yeah, I've seen that you can determine the batch size to send out. My batch size right now is 150 in a loop because all data has to be encrypted and stored encrypted until it is sent out. Too big of a batch size would ocupy too much memory and processing. My server is huffing and puffing as it is with this batch size (I'm working on this too with multithreading updates). 

But eventually it quicks with this issue after 10 batches this size. 

I'm going to try closing the body as mentioned above with `$response->getBody()->close();
` [at the end of this section](https://github.com/web-push-libs/web-push-php/blob/c92de14d511e1ab7a316d8404678e4acf03c3262/src/WebPush.php#L174) and try again. 

If it works then I'll do a pull request to that library to get that fixed. Hope it does!

**Comment by i-allan** [comment]
I am using laravel queue, and in background, i am sending around 1k hits to the server. after certain period i get this issue. any help here?

**Comment by YOzaz** [comment]
Confirmed, happening for me as well.

**Comment by darkot013** [comment]
Linux server have "open file" limit in global and per process.
 - Global -> **/proc/sys/fs/file-max**
 - Per process -> **ulimit -Sn**  -> default 1024

You can check (just in case) you'r global number, but in most cases server configuration not a problem here.

-----

When you make request (TCP connection) you making 1 open file in process directory (/proc/{pid}/fd/) and this is normal behavior. My problem here #1977 (also probably problem with your script) is because I never close that TCP connection and after 1k+ request I hit limit "open file per process" (1024) and server kill that command.

By default, TCP connection after request done going to status "CLOSE_WAIT" and waiting that process be done. Problem with this is that because every TCP connection which is in that status have one open file and you hit limit very fast. TCP connection never close (if you don't tell them) and this is totally normal because can be reuse and in that way you speed up little bit all requests. In my cases, when I close TCP connection after every request I slow down every request for about 30%, but this is necessary when you make lots request with one process (with Guzzle Pool).

How you can test this theory?
1. Start command which have Guzzle Pool system
2. Get process id (pid)
3. Track how many open files you have for that process -> "**ls /proc/{pid}/fd | wc -l**"
4. Track how many TCP connection you have -> **"lsof -i -a -p {pid}"** 

----

For you @i-allan, @TKafassis cases, if you set in Guzzle Pool "concurrency" more then 1000, you need to increase "open files per process" because default limit is 1024 and when you make 1k request in some moment you have 1k open files and hit server limit... 

You have 2 solution:
 1. Reduce number of concurrency requests
 2. Increase number open files per process (https://www.cyberciti.biz/faq/linux-increase-the-maximum-number-of-open-files/)

In both cases you need change little bit Guzzle configuration and close TCP connection after every request (#1977).

**Comment by pratik-joshi-cactus** [comment]
@darkot013  , for this solution : Reduce number of concurrency requests, we tried your solution but didn't work. Could you please check the comments : https://github.com/guzzle/guzzle/issues/1977#issuecomment-351765309

**Comment by darkot013** [comment]
@pratik-joshi-cactus, can you send my:

 - Which OS?
 - Guzzle Pool and Client configuration which you set?
 - "cat /proc/sys/fs/file-max"
 - How match RAM you have on that server?
 - "ulimit -Sn"

**Comment by sagikazarmark** [comment]
Recently I had an interesting bug in another piece of software (unrelated to Guzzle) and I found that destructors are not always called immediately using apache+mod_php combo. So one potential fix can be manually closing the underlying streams (when using the stream handler).

I would also add a question: which handler do you use?

**Comment by pablorsk** [comment]
How can you up file open limit :wrench: :
```bash
ulimit -n 10000
```

Solves problem with `phpunit` or/and `phpdbg` and `Warning:  Uncaught ErrorException: require([..file]): failed to open stream: Too many open files in  [...]`

**Comment by shtrih** [comment]
The problem occurs when you didn't use `handler` option in `\GuzzleHttp\Client` constructor and recreate `Client` instance for new requests (when use DI for example). So it automatically create default handler, which instantiating with hardcoded `3` idle curl connections:
https://github.com/guzzle/guzzle/blob/9133fe03c4df51c50303ddc61612c0b66b487c12/src/Client.php#L64-L65
https://github.com/guzzle/guzzle/blob/9133fe03c4df51c50303ddc61612c0b66b487c12/src/HandlerStack.php#L38-L40
https://github.com/guzzle/guzzle/blob/9133fe03c4df51c50303ddc61612c0b66b487c12/src/functions.php#L103-L107
`new CurlFactory(3)`
https://github.com/guzzle/guzzle/blob/9133fe03c4df51c50303ddc61612c0b66b487c12/src/Handler/CurlHandler.php#L26-L30

So if you send request once as result it doesn't invoke `curl_close` on finishing request and you have min 1 unclosed connection.
https://github.com/guzzle/guzzle/blob/9133fe03c4df51c50303ddc61612c0b66b487c12/src/Handler/CurlHandler.php#L43
https://github.com/guzzle/guzzle/blob/9133fe03c4df51c50303ddc61612c0b66b487c12/src/Handler/CurlFactory.php#L106
https://github.com/guzzle/guzzle/blob/9133fe03c4df51c50303ddc61612c0b66b487c12/src/Handler/CurlFactory.php#L66-L79

There is no problem in most cases but in long term cli script you SHOULD reuse `Client` instance to avoid `Too many open files` error. 

I'm using Laravel, so I used Dependency Injection to create instance of `Client` and encountered this problem. 
In my case solution is [binding a singleton](https://laravel.com/docs/5.6/container#binding-basics):
```php
<?php

namespace App\Providers;

use App\Remote\GuzzleClientWrapper\Client;
use Illuminate\Support\ServiceProvider;

class AppServiceProvider extends ServiceProvider
{
    /**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        $this->app->singleton(Client::class, function () {
            return new Client;
        });
    }
}
```

**Comment by tylik1** [comment]
Any update on this issue? It is impossible to use the lib.
 I'm not sending tons of requests, just testing oauth, manually doing requests, one at a time.
 It seems that when creating new Client, it doesn't shutdown itself, when script ends. Is there a way to shut it down manually, when data is received?

**Comment by thumbtech** [comment]
I am assuming here that @shtrih has identified the issue and that this qualifies as a bug(?). I have a CakePHP CLI running flysystem depending on aws/aws-sdk-php depending on guzzle ... not sure I would have the opportunity to override the misuse of the handle-less guzzle clients in the stack. But shouldn't the default handler behave better?

(One other note that will either cloud or clarify: Rolling back from PHP 7.* to 5.6 on the same underlying linux instance eliminated this error for me!?!)

**Comment by arderyp** [comment]
do any maintainers want to chime in here?  I am also experiencing this.  Is it a bug, or the result of us following poor implementation.

What I'm basically doing is:

```
class SomeClass
{
    $client = ...;
    $promises = [];

    private function someFunction(int $batchSize = 0)
    {
        foreach ($list as $item) {
            $this->promises[] = $this->client->requestAsync($this->method, $path, $options)->then(
                function ($response) {
                    return $this->processResponse($response);
                },
                function (RequestException $exception) {
                     return $this->handleRequestException($exception);
                }
            );
        }

        // Process promises by batch of 20, since we have no clue how big $list is at any given run
        $offset = 0;
        $batchSize = $batchSize :? count($this->promises);  // if no arg provided, process all promises simultaneously
        $promiseCount = count($this->promises);
        while ($offset < $promiseCount) {
            $promiseSubset = array_slice($this->promises, $offset, $batchSize, true);
            $results = Promise\unwrap($set);
            foreach ($results as $key => $result) {
                $this->data[$key] = $result;
            }
            $offset += $batchSize;
        }
    }
```

The code works perfectly in most of my use cases, where I don't set a `$batchSize`, but rather let all of my requests process at once.  This can handle pretty large numbers of requests/responses at once.  However, I just created a new use case where the processing load is just too big, and it resulted in the error this issue is about.  So, I decided to try to batch the number of promises processed by passing in `$batchSize = 20`, and the same error results.  Even if I pass in `$batchSize = 1` it still results in the same error.  Basically, it looks like, despite the while loop, all of my promises are getting processed simultaneously.  Am I just misunderstanding how `unwrap()` and promises work?

**Comment by arderyp** [comment]
@sagikazarmark, do you mind explaining the Guzzle 6 method of doing what you proposed in this comment: https://github.com/guzzle/guzzle/issues/1927#issuecomment-352522542

**Comment by mariuskurgonas** [comment]
Ran into a similar issue that the connection was kept alive. With the latest Guzzle version, solved it by:

'headers' => [
        'Connection' => 'close',
    ]

Then running the Guzzle request again with debug enabled - it reported connection to be closed and recognised the header.

**Comment by arderyp** [comment]
@mariuskurgonas thanks for the feedback.  Someone else also suggested that, but alas the header trick doesn't work for my use case.  Can you provide more context about the requests you were making, and how you were making them?  (how many requests, were they all issued by the same Client object in the same PHP session, were they [concurrent](http://docs.guzzlephp.org/en/stable/quickstart.html#concurrent-requests) requests, or regular old linear requests, how long does your process take to run, etc)

@LiamKarlMitchell and @shawnhind, what confuses you about my comment?  I'd be happy to clarify as best I can.

Has anyone commenting in this issue who encountered this problem _using [concurrent](http://docs.guzzlephp.org/en/stable/quickstart.html#concurrent-requests) requests_ found a solution that worked for them?  If so, do you mind sharing some code?

**Comment by michaelbutler** [comment]
@arderyp indeed creating the custom CurlHandler object with `maxHandles` set to 1 also "resolved" the issue for me; and like you said it just makes our CLI script a lot slower -- since we're no longer reusing curl connections. I used Xdebug to confirm that when I set `maxHandles` to 1, `curl_close` is called after each request and the file descriptors does not grow continously.

When setting `maxHandles` to 3, I confirmed that `curl_close` is _never_
called but also confirmed that `curl_init()` is _not_ called after the initial create! GuzzleHttp is correctly popping off the handle from the handles array, and resetting it using `curl_reset` and `curl_setopt_array` it for the next request. Could this maybe be a cURL library bug or similar? Maybe we can repro this in a simple docker container using various cURL, PHP, and Guzzle versions.

**Comment by arderyp** [comment]
hey @michaelbutler, thanks for that indept info!

I think we can both agree that the `maxHandles=1` approach isn't a real solution, since its not scalable (and this is, in and of itself, an issue about an error resulting from scaling large numbers of requests).

So does it seem to you that the root problem is that `curl_close` is never called when using a number greater than 1 for `maxHandles`? It seemed to me (as noted in my comment [here](https://github.com/guzzle/guzzle/issues/1927#issuecomment-431467861)) that setting `maxHandles` to a number greater than 1 didn't speed up my process at all.  I would have thought that, for example, a session that issues 100 requests would run 50 times faster when using `maxHandles=50` when compared to `maxHandles=1`.  In other words, I thought that `maxHandles` was meant to determine the number of maximum concurrent request sessions/handlers that could be open/processing at a given time.  Is that understanding incorrect?

Whatever the problem is, you seem to be onto something that I couldn't get at.  I _really_ appreciate you looking into this.

**Comment by michaelbutler** [comment]
Making some progress here. I've tested adding `__destruct` to various core guzzle classes relating to cURL -- I'm not exactly sure why they do not have `__destruct` implemented on them.  I found that in a simple test script I made, I do not see the open files issue, and `__destruct` gets called on CurlFactory after every iteration of my test function, since once the local variables go out of scope, all elements should be freed up and destructed. However in my production code, the destructors _do not_ get called until the _script_ exits, and I _am_ seeing file descriptors being held on to. 

If I can pinpoint why the objects are not getting destructed we may have a solution here. Will continue trying to narrow it down between the example script and production script.

**Comment by arderyp** [comment]
> once the local variables go out of scope, all elements should be freed up and destructed

Yes!

Thank you so much for digging into this.  If you want me to test your destructor code in my use case, just provide the code block(s) and tell me which file(s)/class(es) to target.

**Comment by michaelbutler** [comment]
@arderyp So, good and bad news. Would you happen to be using Guzzle Middleware? I finally tracked down the destructors problem: as soon as I commented out my Middleware logic, everything worked as expected, variables got cleared out as well as file descriptors. 

Later, or tomorrow I will port this to the example script to reproduce it.

**Comment by arderyp** [comment]
Hey @michaelbutler, I'm glad you tracked that down.  However, I don't believe I am currently injecting/using any middleware.  Is there some generic way for me to test that its not being injected somewhere I might not be immediately aware of... so way to dump middleware objects from the client?

**Comment by michaelbutler** [comment]
OK, I have at least one possible lead now. I was able to prove that if even just one URL of the entire batch causes an exception (eg. a timeout or even a simple 404 not found status response) it immediately triggers the "bug" where NONE of the class destructors get called, file descriptors stay open, etc. Try this test gist for yourself: https://gist.github.com/michaelbutler/73bf7e52f52224232fcb123dafe4df39

Drop it in the root of any project which has guzzle installed and `vendor/autoload.php` available.

In the batch of paths, one is purposely wrong which causes a 404, and trigger the fail callback. Other than that everything continues working, except the FD count increases the more iterations run, until process exit. 

On the other hand, if you simple comment out the path which generates the 404, all the destructors call normally, and FD count never exceeds a low static number (8 in my testing).

I still am not able to reproduce in an example script when _all_ the URLs result in success, but this sounds like a problem that needs to be dealt with anyway.

**Comment by arderyp** [comment]
Hey @michaelbutler,  I wonder if using `->then(...)` [chained functions off of asyncronous promises](http://docs.guzzlephp.org/en/stable/faq.html#can-guzzle-send-asynchronous-requests) might run into the same issue as the middleware issue you mentioned, seeing as it's a custom handler function for processing the response.  I _am_ using that for my requests... so maybe that's why I am seeing the problem too.

I have yet to determine if any of my requests are unsuccessful (404).  I need to actually use `maxHandles` to run the process linearly to extract this info, which will take some time.  I will report back.

**Comment by michaelbutler** [comment]
@arderyp glad you were able to reproduce; based on the original commenter here who was doing 5000 concurrent requests, I figured there was a high likelihood that at least one caused an exception of some kind, and since otherwise the application worked, this might have been overlooked. With that in mind we can begin to track down why a thrown exception causes the connection leak. Maybe it's related to [this issue](https://github.com/guzzle/guzzle/issues/2266) since it mentions something in the EasyHandle is not try/catching properly.

**Comment by arderyp** [comment]
@michaelbutler, for what it is worth, I wrapped the `$easy->createResponse();` call mentioned in https://github.com/guzzle/guzzle/issues/2266 in a try/catch block and that didn't resolve the problem. But, as @not-implemented seems to imply in his OP on https://github.com/guzzle/guzzle/issues/2266, maybe the exception can't be merely caught but has to be handled in some specific fashion to resolve the issue.

**Comment by michaelbutler** [comment]
OK I think I've reached my limit of this problem, I can't tell exactly where it's happening due to the level and deepness of callbacks and methods, but at least we have a fairly simple repro script. I'm going to open a new issue with the repro script for more visibility. I don't even know if the problem is specific to the guzzle/promises library or not.

**Comment by michaelbutler** [comment]
Sorry, I spoke too soon. Going over one final check I forgot to test this with out the `'sink' => '/dev/null'` option, which it turns out we are using in our project to ignore the response bodies. When I comment this part out, the FD leak does not occur (but then, more memory will be used since the response data needs to be stored _somewhere_). Is anyone here using `/dev/null` or other custom `sink`?

**Comment by arderyp** [comment]
@michaelbutler, thank you for your persistence.

I am not using the `sink` option, but I am seeing this issue and have determined that I have multiple 404s in the promise request chain.  You last comment makes it sound as if 404s in the request chain are not enough to produce this problem, but that a custom `sink` but also be implemented, but unfortunately my use case doesn't corroborate that.  Maybe I am misunderstanding your message.

When I run your linked script as is, I see the FD fount increment from `5 --> 19 --> 33 --> 47 --> 61`, and when I comment out the `sink` option, the FD count still increments, but this time from `5 --> 18 --> 31 --> 44 --> 57`.  So the `sink` seems to have some effect, but there seems to be something else--or possibly multiple other factors--at play.

**Comment by michaelbutler** [comment]
Hi all I opened a new issue with a repro test case: https://github.com/guzzle/guzzle/issues/2289

**Comment by arderyp** [comment]
thanks @michaelbutler !!

**Comment by michaelbutler** [comment]
Hello again all, I have finally gotten to the bottom of this and have a workaround. It involves creating a CurlFactory class override in your own repository that extends Guzzle's CurlFactory. A custom method `cleanup` is added that you have to call periodically -- ideally when your inner code is completely finished making requests and using the response(s). In my testing, this never increases the open file descriptor count and also seems to be memory efficient and also fast (does requests in parallel asynchronously)

See this updated [gist file](https://github.tumblr.net/gist/mbutler/06d94089e391d3343161cc2b2dd4fe40) which includes the example class override.

I believe the core problem is that `curl_close` is never called on the handles in CurlFactory even if your own functions return and their local variables not needed anymore. By adding the `cleanup` method we loop through and manually close these open connections.

**Comment by arderyp** [comment]
thanks @michaelbutler.  I want to test this, but your gist link is broken

**Comment by arderyp** [comment]
I'm guessing this is what you meant to post: https://gist.github.com/michaelbutler/14195fbf91a525c5af0cdcbf819095eb

**Comment by michaelbutler** [comment]
Apologies, link updated.

**Comment by arderyp** [comment]
@michaelbutler, when I updated from php 7.2 to 7.4,  [your gist](https://gist.github.com/michaelbutler/73bf7e52f52224232fcb123dafe4df39) just worked.

In addition to updating php, I've also updated guzzle from `6.3.3` to `6.5.5`.

Regardless, the approach I outlined [here](https://github.com/guzzle/guzzle/issues/1927#issuecomment-429488906) is still causing this error, despite this environment producing no errors with your gist.

**Comment by arderyp** [comment]
@michaelbutler , it seems to be a simple issue of volume.  All I did was take your gist and radically increased the number of calls: https://gist.github.com/arderyp/eb50a03e033b40de9b83c0cec44e9605

So, it seems like some post 7.2 fix in PHP resolved the memory leak issue, but the core issue remains: if you provide too many requests to process concurrently, you'll reach this error:

```
** IN BEGIN: OPEN FDs 4
PHP Warning:  include(/home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Exception/ConnectException.php): failed to open stream: Too many open files in /home/arderyp/code/TEST/vendor/composer/ClassLoader.php on line 444
PHP Stack trace:
PHP   1. {main}() /home/arderyp/code/TEST/guzzletest_phil.php:0
PHP   2. run_outer_loop() /home/arderyp/code/TEST/guzzletest_phil.php:1742
PHP   3. run_script_chunk() /home/arderyp/code/TEST/guzzletest_phil.php:1733
PHP   4. GuzzleHttp\Promise\Promise->wait() /home/arderyp/code/TEST/guzzletest_phil.php:1722
PHP   5. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:62
PHP   6. GuzzleHttp\Promise\Promise->invokeWaitList() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:225
PHP   7. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:267
PHP   8. GuzzleHttp\Promise\Promise->invokeWaitFn() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:223
PHP   9. GuzzleHttp\Promise\EachPromise->GuzzleHttp\Promise\{closure:/home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/EachPromise.php:90-106}() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:246
PHP  10. GuzzleHttp\Promise\Promise->wait() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/EachPromise.php:101
PHP  11. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:62
PHP  12. GuzzleHttp\Promise\Promise->invokeWaitList() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:225
PHP  13. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:267
PHP  14. GuzzleHttp\Promise\Promise->invokeWaitFn() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:223
PHP  15. GuzzleHttp\Handler\CurlMultiHandler->execute() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:246
PHP  16. GuzzleHttp\Handler\CurlMultiHandler->tick() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:145
PHP  17. GuzzleHttp\Handler\CurlMultiHandler->processMessages() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:130
PHP  18. GuzzleHttp\Handler\CurlFactory::finish() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:201
PHP  19. GuzzleHttp\Handler\CurlFactory::finishError() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:105
PHP  20. GuzzleHttp\Handler\CurlFactory::createRejection() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:155
PHP  21. spl_autoload_call() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:200
PHP  22. Composer\Autoload\ClassLoader->loadClass() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:200
PHP  23. Composer\Autoload\includeFile() /home/arderyp/code/TEST/vendor/composer/ClassLoader.php:322

Warning: include(/home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Exception/ConnectException.php): failed to open stream: Too many open files in /home/arderyp/code/TEST/vendor/composer/ClassLoader.php on line 444

Call Stack:
    0.0008     641400   1. {main}() /home/arderyp/code/TEST/guzzletest_phil.php:0
    0.1050    1043608   2. run_outer_loop() /home/arderyp/code/TEST/guzzletest_phil.php:1742
    0.1050    1043608   3. run_script_chunk() /home/arderyp/code/TEST/guzzletest_phil.php:1733
    0.5385   34202672   4. GuzzleHttp\Promise\Promise->wait() /home/arderyp/code/TEST/guzzletest_phil.php:1722
    0.5385   34202672   5. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:62
    0.5385   34202672   6. GuzzleHttp\Promise\Promise->invokeWaitList() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:225
    0.5385   34202672   7. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:267
    0.5385   34202672   8. GuzzleHttp\Promise\Promise->invokeWaitFn() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:223
    0.5385   34202672   9. GuzzleHttp\Promise\EachPromise->GuzzleHttp\Promise\{closure:/home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/EachPromise.php:90-106}() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:246
    0.5385   34202704  10. GuzzleHttp\Promise\Promise->wait() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/EachPromise.php:101
    0.5385   34202704  11. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:62
    0.5385   34202704  12. GuzzleHttp\Promise\Promise->invokeWaitList() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:225
    0.5385   34202704  13. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:267
    0.5385   34202704  14. GuzzleHttp\Promise\Promise->invokeWaitFn() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:223
    0.5385   34202704  15. GuzzleHttp\Handler\CurlMultiHandler->execute() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:246
    0.7463   34211896  16. GuzzleHttp\Handler\CurlMultiHandler->tick() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:145
    0.8413   34211896  17. GuzzleHttp\Handler\CurlMultiHandler->processMessages() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:130
    0.8415   34212384  18. GuzzleHttp\Handler\CurlFactory::finish() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:201
    0.8415   34212384  19. GuzzleHttp\Handler\CurlFactory::finishError() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:105
    0.8416   34215560  20. GuzzleHttp\Handler\CurlFactory::createRejection() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:155
    0.8416   34215456  21. spl_autoload_call() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:200
    0.8416   34215520  22. Composer\Autoload\ClassLoader->loadClass() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:200
    0.8416   34215680  23. Composer\Autoload\includeFile() /home/arderyp/code/TEST/vendor/composer/ClassLoader.php:322

PHP Warning:  include(): Failed opening '/home/arderyp/code/TEST/vendor/composer/../guzzlehttp/guzzle/src/Exception/ConnectException.php' for inclusion (include_path='.:/usr/share/php') in /home/arderyp/code/TEST/vendor/composer/ClassLoader.php on line 444
PHP Stack trace:
PHP   1. {main}() /home/arderyp/code/TEST/guzzletest_phil.php:0
PHP   2. run_outer_loop() /home/arderyp/code/TEST/guzzletest_phil.php:1742
PHP   3. run_script_chunk() /home/arderyp/code/TEST/guzzletest_phil.php:1733
PHP   4. GuzzleHttp\Promise\Promise->wait() /home/arderyp/code/TEST/guzzletest_phil.php:1722
PHP   5. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:62
PHP   6. GuzzleHttp\Promise\Promise->invokeWaitList() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:225
PHP   7. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:267
PHP   8. GuzzleHttp\Promise\Promise->invokeWaitFn() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:223
PHP   9. GuzzleHttp\Promise\EachPromise->GuzzleHttp\Promise\{closure:/home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/EachPromise.php:90-106}() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:246
PHP  10. GuzzleHttp\Promise\Promise->wait() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/EachPromise.php:101
PHP  11. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:62
PHP  12. GuzzleHttp\Promise\Promise->invokeWaitList() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:225
PHP  13. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:267
PHP  14. GuzzleHttp\Promise\Promise->invokeWaitFn() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:223
PHP  15. GuzzleHttp\Handler\CurlMultiHandler->execute() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:246
PHP  16. GuzzleHttp\Handler\CurlMultiHandler->tick() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:145
PHP  17. GuzzleHttp\Handler\CurlMultiHandler->processMessages() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:130
PHP  18. GuzzleHttp\Handler\CurlFactory::finish() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:201
PHP  19. GuzzleHttp\Handler\CurlFactory::finishError() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:105
PHP  20. GuzzleHttp\Handler\CurlFactory::createRejection() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:155
PHP  21. spl_autoload_call() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:200
PHP  22. Composer\Autoload\ClassLoader->loadClass() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:200
PHP  23. Composer\Autoload\includeFile() /home/arderyp/code/TEST/vendor/composer/ClassLoader.php:322

Warning: include(): Failed opening '/home/arderyp/code/TEST/vendor/composer/../guzzlehttp/guzzle/src/Exception/ConnectException.php' for inclusion (include_path='.:/usr/share/php') in /home/arderyp/code/TEST/vendor/composer/ClassLoader.php on line 444

Call Stack:
    0.0008     641400   1. {main}() /home/arderyp/code/TEST/guzzletest_phil.php:0
    0.1050    1043608   2. run_outer_loop() /home/arderyp/code/TEST/guzzletest_phil.php:1742
    0.1050    1043608   3. run_script_chunk() /home/arderyp/code/TEST/guzzletest_phil.php:1733
    0.5385   34202672   4. GuzzleHttp\Promise\Promise->wait() /home/arderyp/code/TEST/guzzletest_phil.php:1722
    0.5385   34202672   5. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:62
    0.5385   34202672   6. GuzzleHttp\Promise\Promise->invokeWaitList() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:225
    0.5385   34202672   7. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:267
    0.5385   34202672   8. GuzzleHttp\Promise\Promise->invokeWaitFn() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:223
    0.5385   34202672   9. GuzzleHttp\Promise\EachPromise->GuzzleHttp\Promise\{closure:/home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/EachPromise.php:90-106}() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:246
    0.5385   34202704  10. GuzzleHttp\Promise\Promise->wait() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/EachPromise.php:101
    0.5385   34202704  11. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:62
    0.5385   34202704  12. GuzzleHttp\Promise\Promise->invokeWaitList() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:225
    0.5385   34202704  13. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:267
    0.5385   34202704  14. GuzzleHttp\Promise\Promise->invokeWaitFn() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:223
    0.5385   34202704  15. GuzzleHttp\Handler\CurlMultiHandler->execute() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:246
    0.7463   34211896  16. GuzzleHttp\Handler\CurlMultiHandler->tick() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:145
    0.8413   34211896  17. GuzzleHttp\Handler\CurlMultiHandler->processMessages() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:130
    0.8415   34212384  18. GuzzleHttp\Handler\CurlFactory::finish() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:201
    0.8415   34212384  19. GuzzleHttp\Handler\CurlFactory::finishError() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:105
    0.8416   34215560  20. GuzzleHttp\Handler\CurlFactory::createRejection() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:155
    0.8416   34215456  21. spl_autoload_call() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:200
    0.8416   34215520  22. Composer\Autoload\ClassLoader->loadClass() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:200
    0.8416   34215680  23. Composer\Autoload\includeFile() /home/arderyp/code/TEST/vendor/composer/ClassLoader.php:322

PHP Fatal error:  Uncaught Error: Class 'GuzzleHttp\Exception\ConnectException' not found in /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:200
Stack trace:
#0 /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(155): GuzzleHttp\Handler\CurlFactory::createRejection()
#1 /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(105): GuzzleHttp\Handler\CurlFactory::finishError()
#2 /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(201): GuzzleHttp\Handler\CurlFactory::finish()
#3 /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(130): GuzzleHttp\Handler\CurlMultiHandler->processMessages()
#4 /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(145): GuzzleHttp\Handler\CurlMultiHand in /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php on line 200

Fatal error: Uncaught Error: Class 'GuzzleHttp\Exception\ConnectException' not found in /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php on line 200

Error: Class 'GuzzleHttp\Exception\ConnectException' not found in /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php on line 200

Call Stack:
    0.0008     641400   1. {main}() /home/arderyp/code/TEST/guzzletest_phil.php:0
    0.1050    1043608   2. run_outer_loop() /home/arderyp/code/TEST/guzzletest_phil.php:1742
    0.1050    1043608   3. run_script_chunk() /home/arderyp/code/TEST/guzzletest_phil.php:1733
    0.5385   34202672   4. GuzzleHttp\Promise\Promise->wait() /home/arderyp/code/TEST/guzzletest_phil.php:1722
    0.5385   34202672   5. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:62
    0.5385   34202672   6. GuzzleHttp\Promise\Promise->invokeWaitList() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:225
    0.5385   34202672   7. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:267
    0.5385   34202672   8. GuzzleHttp\Promise\Promise->invokeWaitFn() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:223
    0.5385   34202672   9. GuzzleHttp\Promise\EachPromise->GuzzleHttp\Promise\{closure:/home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/EachPromise.php:90-106}() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:246
    0.5385   34202704  10. GuzzleHttp\Promise\Promise->wait() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/EachPromise.php:101
    0.5385   34202704  11. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:62
    0.5385   34202704  12. GuzzleHttp\Promise\Promise->invokeWaitList() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:225
    0.5385   34202704  13. GuzzleHttp\Promise\Promise->waitIfPending() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:267
    0.5385   34202704  14. GuzzleHttp\Promise\Promise->invokeWaitFn() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:223
    0.5385   34202704  15. GuzzleHttp\Handler\CurlMultiHandler->execute() /home/arderyp/code/TEST/vendor/guzzlehttp/promises/src/Promise.php:246
    0.7463   34211896  16. GuzzleHttp\Handler\CurlMultiHandler->tick() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:145
    0.8413   34211896  17. GuzzleHttp\Handler\CurlMultiHandler->processMessages() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:130
    0.8415   34212384  18. GuzzleHttp\Handler\CurlFactory::finish() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php:201
    0.8415   34212384  19. GuzzleHttp\Handler\CurlFactory::finishError() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:105
    0.8416   34215560  20. GuzzleHttp\Handler\CurlFactory::createRejection() /home/arderyp/code/TEST/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:155
```

In a less extreme, real-world context, I issues nearly 500 requests concurrently without issue, and dumped the FD count in a way similar to your gist, and it seems to hold consistently and not leak out of control as was happening on php 7.2 with your gist.

**Comment by Nyholm** [comment]
I’m closing this as it is not a bug related to guzzle.
